Normed real vector spaces (#1704)
UniMath/SymmetryBookFormalization: src/linear-algebra/seminormed-real-vector-spaces.lagda.md

A seminormed space has an induced
[pseudometric structure](metric-spaces.pseudometric-spaces.md) defined by the
neighborhood relation that `v` and `w` are in an `ε`-neighborhood of each other
if `p(v - w) ≤ ε`.

## Definition

```agda
module _
  {l1 l2 : Level}
  (V : ℝ-Vector-Space l1 l2)
  (p : type-ℝ-Vector-Space V → ℝ l1)
  where

  is-triangular-prop-seminorm-ℝ-Vector-Space : Prop (l1 ⊔ l2)
  is-triangular-prop-seminorm-ℝ-Vector-Space =
    Π-Prop
      ( type-ℝ-Vector-Space V)
      ( λ x →
        Π-Prop
          ( type-ℝ-Vector-Space V)
          ( λ y → leq-prop-ℝ (p (add-ℝ-Vector-Space V x y)) (p x +ℝ p y)))

  is-triangular-seminorm-ℝ-Vector-Space : UU (l1 ⊔ l2)
  is-triangular-seminorm-ℝ-Vector-Space =
    type-Prop is-triangular-prop-seminorm-ℝ-Vector-Space

  is-absolutely-homogeneous-prop-seminorm-ℝ-Vector-Space : Prop (lsuc l1 ⊔ l2)
  is-absolutely-homogeneous-prop-seminorm-ℝ-Vector-Space =
    Π-Prop
      ( ℝ l1)
      ( λ c →
        Π-Prop
          ( type-ℝ-Vector-Space V)
          ( λ x →
            Id-Prop
              ( ℝ-Set l1)
              ( p (mul-ℝ-Vector-Space V c x))

>               ( abs-ℝ c *ℝ p x)))

Author: Yes, and?  That's the only one that applies to seminorms, and the name reflects that.  Wikipedia describes this condition as [absolute homogeneity](https://en.wikipedia.org/wiki/Seminorm) with no qualifier.

>               ( abs-ℝ c *ℝ p x)))

Reviewer: For the record I had only heard of the concept of homogeneity in the context of homogeneous polynomials, and the connection to the condition you've written here is not very obvious to me.

>               ( abs-ℝ c *ℝ p x)))

Author: I'm not really sure I understand it myself, I'll admit, only that these are the terms Wikipedia uses that appear to be standard.

>               ( abs-ℝ c *ℝ p x)))

Reviewer: I suggested a more elaborate explanation here https://github.com/UniMath/agda-unimath/pull/1704#discussion_r2566401767

    zero-Seminormed-ℝ-Vector-Space
  left-zero-law-mul-Seminormed-ℝ-Vector-Space =
    left-zero-law-mul-ℝ-Vector-Space vector-space-Seminormed-ℝ-Vector-Space

  mul-neg-one-Seminormed-ℝ-Vector-Space :
    (v : type-Seminormed-ℝ-Vector-Space) →
    mul-Seminormed-ℝ-Vector-Space (neg-ℝ (raise-ℝ l1 one-ℝ)) v ＝
    neg-Seminormed-ℝ-Vector-Space v
  mul-neg-one-Seminormed-ℝ-Vector-Space =
    mul-neg-one-ℝ-Vector-Space vector-space-Seminormed-ℝ-Vector-Space

  distributive-neg-diff-Seminormed-ℝ-Vector-Space :
    (v w : type-Seminormed-ℝ-Vector-Space) →
    neg-Seminormed-ℝ-Vector-Space (diff-Seminormed-ℝ-Vector-Space v w) ＝
    diff-Seminormed-ℝ-Vector-Space w v
  distributive-neg-diff-Seminormed-ℝ-Vector-Space =
    neg-right-subtraction-Ab
      ( ab-ℝ-Vector-Space vector-space-Seminormed-ℝ-Vector-Space)

  triangular-Seminormed-ℝ-Vector-Space :
    (v w : type-Seminormed-ℝ-Vector-Space) →
    leq-ℝ
      ( map-seminorm-Seminormed-ℝ-Vector-Space
        ( add-Seminormed-ℝ-Vector-Space v w))
      ( map-seminorm-Seminormed-ℝ-Vector-Space v +ℝ
        map-seminorm-Seminormed-ℝ-Vector-Space w)
  triangular-Seminormed-ℝ-Vector-Space =
    pr1 (pr2 seminorm-Seminormed-ℝ-Vector-Space)

>   is-absolutely-homogeneous-Seminormed-ℝ-Vector-Space :

Reviewer: ```suggestion
  is-absolutely-homogeneous-seminorm-Seminormed-ℝ-Vector-Space :
```

open import real-numbers.similarity-real-numbers
open import real-numbers.strict-inequality-real-numbers
```

</details>

## Idea

A
{{#concept "seminorm" WDID=Q1416088 WD="seminorm" Disambiguation="on a real vector space" Agda=seminorm-ℝ-Vector-Space}}
on a [real vector space](linear-algebra.real-vector-spaces.md) `V` is a
[real](real-numbers.dedekind-real-numbers.md)-valued function `p` on the vector
space such that `p(x + y) ≤ p(x) + p(y)` for all `x` and `y` in `V`, and
`p(c * x) = |c| * p(x)` for all real numbers `c` and `x` in `V`. This latter
condition is referred to as absolute homogeneity, and is related to the more
general concept of
[homogeneous functions](https://en.wikipedia.org/wiki/Homogeneous_function).

These conditions imply that `p(0) = 0` and that `p` is nonnegative.

A real vector space equipped with such a seminorm is called a
{{#concept "seminormed space" WD="seminormed space" WDID=Q63793693 Agda=Seminormed-ℝ-Vector-Space}}.
A seminormed space has an induced
[pseudometric structure](metric-spaces.pseudometric-spaces.md) defined by the
neighborhood relation that `v` and `w` are in an `ε`-neighborhood of each other

> if `p(v - w) ≤ ε`.

Reviewer: ```suggestion
A
{{#concept "seminorm" WDID=Q1416088 WD="seminorm" Disambiguation="on a real vector space" Agda=seminorm-ℝ-Vector-Space}}
on a [real vector space](linear-algebra.real-vector-spaces.md) `V` is a
[real](real-numbers.dedekind-real-numbers.md)-valued function `p` on the vector
space which is

- **Triangular.** `p(x + y) ≤ p(x) + p(y)` for all `x` and `y` in `V`, and
- **Absolutely homogeneous.** `p(c * x) = |c| * p(x)` for all real numbers `c`
  and `x` in `V`.

These conditions together imply that `p` is extensional and semidefinite, in
other words, that `p(0) = 0` and `p(x) ≥ 0`.

A real vector space equipped with such a seminorm is called a
{{#concept "seminormed space" WD="seminormed space" WDID=Q63793693 Agda=Seminormed-ℝ-Vector-Space}}.
A seminormed space has an induced
[pseudometric structure](metric-spaces.pseudometric-spaces.md) defined by the
neighborhood relation that `v` and `w` are in an `ε`-neighborhood of each other
if `p(v - w) ≤ ε`.

**Terminology.** The term _absolute homogeneity_ comes from the more general
concept of
[homogeneous functions](https://en.wikipedia.org/wiki/Homogeneous_function). A
multivariable function `f` on real vector spaces is said to be _homogeneous of
degree `k`_ if for every scalar `s` we have that
`f(sx₀,sx₁,…,sxₙ) = sᵏf(x₀,x₁,…,xₙ)`. In particular, you may note that any
homogeneous bilinear form must be homogeneous of degree one.
```

open import real-numbers.inequality-real-numbers
open import real-numbers.multiplication-positive-real-numbers
open import real-numbers.multiplication-real-numbers
open import real-numbers.negation-real-numbers
open import real-numbers.nonnegative-real-numbers
open import real-numbers.positive-real-numbers
open import real-numbers.raising-universe-levels-real-numbers
open import real-numbers.rational-real-numbers
open import real-numbers.saturation-inequality-nonnegative-real-numbers
open import real-numbers.similarity-real-numbers
open import real-numbers.strict-inequality-real-numbers
```

</details>

## Idea

A
{{#concept "seminorm" WDID=Q1416088 WD="seminorm" Disambiguation="on a real vector space" Agda=seminorm-ℝ-Vector-Space}}
on a [real vector space](linear-algebra.real-vector-spaces.md) `V` is a
[real](real-numbers.dedekind-real-numbers.md)-valued function `p` on the vector
space which is

- **Triangular.** `p(x + y) ≤ p(x) + p(y)` for all `x` and `y` in `V`, and
- **Absolutely homogeneous.** `p(c * x) = |c| * p(x)` for all real numbers `c`
  and `x` in `V`.

These conditions together imply that `p` is extensional and semidefinite, in

> other words, that `p(0) = 0` and `p(x) ≥ 0`.

Reviewer: Whoops, a mistake snuck in
```suggestion
These conditions together imply that `p` is semidefinite and sends the zero
vector to zero, i.e., `p(x) ≥ 0` and `p(0) = 0`.
```

  {l1 l2 : Level}
  (V : Seminormed-ℝ-Vector-Space l1 l2)
  where

  abstract
    triangular-dist-Seminormed-ℝ-Vector-Space :
      (v w x : type-Seminormed-ℝ-Vector-Space V) →
      leq-ℝ
        ( dist-Seminormed-ℝ-Vector-Space V v x)
        ( dist-Seminormed-ℝ-Vector-Space V v w +ℝ
          dist-Seminormed-ℝ-Vector-Space V w x)
    triangular-dist-Seminormed-ℝ-Vector-Space v w x =
      let
        open inequality-reasoning-Large-Poset ℝ-Large-Poset
      in
        chain-of-inequalities
          dist-Seminormed-ℝ-Vector-Space V v x
          ≤ map-seminorm-Seminormed-ℝ-Vector-Space
            ( V)
            ( add-Seminormed-ℝ-Vector-Space
              ( V)
              ( diff-Seminormed-ℝ-Vector-Space V v w)
              ( diff-Seminormed-ℝ-Vector-Space V w x))
            by
              leq-eq-ℝ
                ( ap
                  ( map-seminorm-Seminormed-ℝ-Vector-Space V)
                  ( inv (add-diff-Seminormed-ℝ-Vector-Space V v w x)))
          ≤ ( dist-Seminormed-ℝ-Vector-Space V v w) +ℝ
            ( dist-Seminormed-ℝ-Vector-Space V w x)
            by triangular-seminorm-Seminormed-ℝ-Vector-Space V _ _
```

### The seminorm of a vector is nonnegative

> 

Reviewer: ```suggestion
### The seminorm of a vector is nonnegative

In other words, seminorms are semidefinite.

```

    diff-Seminormed-ℝ-Vector-Space w v
  distributive-neg-diff-Seminormed-ℝ-Vector-Space =
    neg-right-subtraction-Ab
      ( ab-ℝ-Vector-Space vector-space-Seminormed-ℝ-Vector-Space)

  triangular-Seminormed-ℝ-Vector-Space :
    (v w : type-Seminormed-ℝ-Vector-Space) →
    leq-ℝ
      ( map-seminorm-Seminormed-ℝ-Vector-Space
        ( add-Seminormed-ℝ-Vector-Space v w))
      ( map-seminorm-Seminormed-ℝ-Vector-Space v +ℝ
        map-seminorm-Seminormed-ℝ-Vector-Space w)
  triangular-Seminormed-ℝ-Vector-Space =
    pr1 (pr2 seminorm-Seminormed-ℝ-Vector-Space)

  is-absolutely-homogeneous-Seminormed-ℝ-Vector-Space :
    (c : ℝ l1) (v : type-Seminormed-ℝ-Vector-Space) →
    map-seminorm-Seminormed-ℝ-Vector-Space
      ( mul-Seminormed-ℝ-Vector-Space c v) ＝
    abs-ℝ c *ℝ map-seminorm-Seminormed-ℝ-Vector-Space v
  is-absolutely-homogeneous-Seminormed-ℝ-Vector-Space =
    pr2 (pr2 seminorm-Seminormed-ℝ-Vector-Space)

  dist-Seminormed-ℝ-Vector-Space :
    type-Seminormed-ℝ-Vector-Space → type-Seminormed-ℝ-Vector-Space → ℝ l1
  dist-Seminormed-ℝ-Vector-Space v w =
    map-seminorm-Seminormed-ℝ-Vector-Space
      ( diff-Seminormed-ℝ-Vector-Space v w)
```

> ### The seminorm of the zero vector in a seminormed real vector space is zero

Reviewer: We're already on the page about seminormed real vector spaces
```suggestion
### The seminorm of the zero vector is zero
```

                ( V)
                ( raise-ℝ l1 zero-ℝ)
                ( zero-Seminormed-ℝ-Vector-Space V))
            by
              sim-eq-ℝ
                ( ap
                  ( map-seminorm-Seminormed-ℝ-Vector-Space V)
                  ( inv (left-zero-law-mul-Seminormed-ℝ-Vector-Space V _)))
          ~ℝ
            ( abs-ℝ (raise-ℝ l1 zero-ℝ)) *ℝ
            ( map-seminorm-Seminormed-ℝ-Vector-Space
              ( V)
              ( zero-Seminormed-ℝ-Vector-Space V))
            by
              sim-eq-ℝ
                ( is-absolutely-homogeneous-Seminormed-ℝ-Vector-Space V _ _)
          ~ℝ
            ( abs-ℝ zero-ℝ) *ℝ
            ( map-seminorm-Seminormed-ℝ-Vector-Space
              ( V)
              ( zero-Seminormed-ℝ-Vector-Space V))
            by
              preserves-sim-right-mul-ℝ _ _ _
                ( preserves-sim-abs-ℝ (sim-raise-ℝ' l1 zero-ℝ))
          ~ℝ
            ( zero-ℝ) *ℝ
            ( map-seminorm-Seminormed-ℝ-Vector-Space
              ( V)
              ( zero-Seminormed-ℝ-Vector-Space V))
            by sim-eq-ℝ (ap-mul-ℝ abs-zero-ℝ refl)
          ~ℝ zero-ℝ
            by left-zero-law-mul-ℝ _
          ~ℝ raise-ℝ l1 zero-ℝ

>             by sim-raise-ℝ l1 zero-ℝ)

Reviewer: oh, alternatively you could register `left-raise-zero-law-mul-ℝ` as the appropriate equality I suppose, and then make this into a equality reasoning argument instead, using  `eq-sim-ℝ` in that one step.

          ( positive-real-ℕ⁺ (2 , λ ()))
          ( zero-ℝ)
          ( map-seminorm-Seminormed-ℝ-Vector-Space V v)
          ( chain-of-inequalities
              real-ℝ⁺ (positive-real-ℕ⁺ (2 , (λ ()))) *ℝ zero-ℝ
              ≤ zero-ℝ
                by leq-sim-ℝ (right-zero-law-mul-ℝ _)
              ≤ raise-ℝ l1 zero-ℝ
                by leq-sim-ℝ (sim-raise-ℝ l1 zero-ℝ)
              ≤ dist-Seminormed-ℝ-Vector-Space V v v
                by
                  leq-eq-ℝ
                    ( inv (is-zero-diagonal-dist-Seminormed-ℝ-Vector-Space V v))
              ≤ ( map-seminorm-Seminormed-ℝ-Vector-Space V v) +ℝ
                ( map-seminorm-Seminormed-ℝ-Vector-Space
                  ( V)
                  ( neg-Seminormed-ℝ-Vector-Space V v))
                by triangular-Seminormed-ℝ-Vector-Space V _ _
              ≤ ( map-seminorm-Seminormed-ℝ-Vector-Space V v) +ℝ
                ( map-seminorm-Seminormed-ℝ-Vector-Space V v)
                by
                  leq-eq-ℝ
                    ( ap-add-ℝ
                      ( refl)
                      ( seminorm-neg-Seminormed-ℝ-Vector-Space V v))
              ≤ real-ℕ 2 *ℝ map-seminorm-Seminormed-ℝ-Vector-Space V v

>                 by leq-eq-ℝ (inv (left-mul-real-ℕ 2 _)))

Reviewer: It should be a lemma that `zero-ℝ` is less than or equal to any nonnegative real, and `real-ℕ 2 *ℝ map-seminorm-Seminormed-ℝ-Vector-Space V v` is nonnegative.
